### ISSUE TITLE
NAT traversal and automatic port forwarding

### ISSUE THREAD

**Issue by layters** [layters/testshop#124]
```
Details
```
In order to establish connections and maintain a decentralized network, nodes need to be able to communicate with each other, even if they are behind NAT (Network Address Translation) routers or a firewall.

By utilizing UPnP and other NAT traversal methods, Neroshop nodes can automate the process of port forwarding, making it easier for users to participate in the network without requiring manual configuration of their routers.

> ELI5: Right now nodes can only send data but cannot receive data because receiving data from their public IP address requires NAT traversal.

```
Additional Information
```
The neroshop [**DHT node**](https://github.com/larteyoh/testshop/blob/main/src/core/protocol/p2p/node.cpp) uses two separate UDP sockets for communication - one for **listening to and responding to requests** via the `Node::run()` function and the other is a temporary socket that is used for **sending queries** to other nodes in the network via the `Node::send_query()` function. I want to know how I can route both the inbound and outbound traffic of the Node using i2p.
```
Files
```
https://github.com/larteyoh/testshop/blob/main/src/core/protocol/p2p/node.cpp
https://github.com/larteyoh/testshop/blob/main/src/daemon/main.cpp

```
Libraries
```
There are a few lightweight NAT traversal libraries available for C/C++ that I've been able to compile so far:

    libnatpmp: a lightweight and portable library for NAT-PMP (NAT Port Mapping Protocol) implementation which I've been able to compile but have no idea how to integrate since I'm just dumb.

    miniupnpc: lightweight library that implements the UPnP (Universal Plug and Play) protocol which is not currently working for me.

> opening a port through UPnP is not reliable for p2p apps. Consider using STUN/TURN/ICE to connect behind NAT

That brings me down to two C/C++ library options:
- [`libnice`](https://github.com/libnice/libnice)
- [`JUICE`](https://github.com/paullouisageneau/libjuice)

I am going with `JUICE` as it has zero dependencies and a very permissive license (🥳), but I will need help with integrating it into the [neroshop daemon](https://github.com/larteyoh/testshop/tree/main/src/daemon) using these [examples](https://github.com/paullouisageneau/libjuice#example) and [**this**](https://github.com/paullouisageneau/libjuice/blob/master/test/notrickle.c) as a reference guide.

```
Related issues
```
[I2P Integration](https://github.com/larteyoh/testshop/issues/7)

**Comment by larteyoh** [comment]
As of May 26 2023, automatic port forwarding now works on UPnP-enabled routers, which should be supported on most modern routers.

Other NAT traversal methods/protocols may be considered in the near future. UPnP should be sufficient for now.

**Comment by larteyoh** [comment]
The miniupnp code in [`main.cpp`](https://github.com/larteyoh/testshop/blob/main/src/daemon/main.cpp#L209) does not work for whatever reason. Any help would be appreciated!
